pkg/pillar/dpcreconciler/linux.go

```diff
@@ -1406,23 +1444,92 @@ func (r *LinuxDpcReconciler) getIntendedWwanConfig(dpc types.DevicePortConfig,
 			physAddress.Interface = port.IfName
 		}
 		network := types.WwanNetworkConfig{
-			// TODO: Username + Password (will be done in the next PR)
-			LogicalLabel:     port.Logicallabel,
-			PhysAddrs:        physAddress,
-			SIMSlot:          accessPoint.SIMSlot,
-			APN:              accessPoint.APN,
-			PreferredPLMNs:   accessPoint.PreferredPLMNs,
-			PreferredRATs:    accessPoint.PreferredRATs,
-			ForbidRoaming:    accessPoint.ForbidRoaming,
-			Proxies:          port.Proxies,
-			Probe:            probeCfg,
-			LocationTracking: locationTracking,
+			LogicalLabel:      port.Logicallabel,
+			PhysAddrs:         physAddress,
+			SIMSlot:           accessPoint.SIMSlot,
+			APN:               accessPoint.APN,
+			AuthProtocol:      accessPoint.AuthProtocol,
+			Username:          decBlock.CellularNetUsername,
+			EncryptedPassword: encPassword,
+			PreferredPLMNs:    accessPoint.PreferredPLMNs,
+			PreferredRATs:     accessPoint.PreferredRATs,
+			ForbidRoaming:     accessPoint.ForbidRoaming,
+			Proxies:           port.Proxies,
+			Probe:             probeCfg,
+			LocationTracking:  locationTracking,
 		}
 		config.Networks = append(config.Networks, network)
 	}
 	return generic.Wwan{Config: config}
 }
 
+func (r *LinuxDpcReconciler) getWwanCredentials(ap *types.CellularAccessPoint) (
+	decBlock types.EncryptionBlock, err error) {
+	if !ap.EncryptedCredentials.IsCipher {
+		return decBlock, nil
+	}
+	decryptAvailable := r.SubControllerCert != nil && r.SubEdgeNodeCert != nil
+	if !decryptAvailable {
+		r.CipherMetrics.RecordFailure(r.Log, types.NotReady)
+		return decBlock, fmt.Errorf(
```

Comment:
Doesn't this cause a failure even if there are no credentials? I think function can just return in that case without even trying to decrypt.

Reply:
Just above it there is:
```
if !ap.EncryptedCredentials.IsCipher {
  return decBlock, nil
}
```

Reply:
But is that IsCipher false when there are no credentials at all? I read it as checking for the (old, unused) cleartext credentials case.

Reply:
Without credentials, the CipherBlock should be left as nil by the controller, otherwise we return error already in the parsing stage: https://github.com/lf-edge/eve/blob/master/pkg/pillar/cmd/zedagent/handlecipherconfig.go#L92-L113
(I have described this in the design document for the zedcloud team)

Reply:
ok